### ISSUE TITLE
[NEXT-1227] `server-only` is not compatible with Server Actions

### ISSUE THREAD

**Issue by abuinitski** [vercel/next.js#50199]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Home Single Language
Binaries:
  Node: 18.15.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant packages:
  next: 13.4.3
  eslint-config-next: 13.4.3
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.0.4
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/summer-leaf-v21228

### To Reproduce

1. Create a typescript file, guarded with `import 'server-only'` (SERVER)
1. Create a client component with `use client` (CLIENT)
2. Create a Server Action with `use server` on file level, e.g. `actions.ts` approach (SA)
3. Import SERVER file into SA, then import SA into the CLIENT

### Describe the Bug

Compilation fails with "You're importing a component that needs server-only. That only works in a Server Component but one of its parents is marked with "use client", so it's a Client Component." error.

E.g. when a file with `import 'server-only'` is in a dependency tree of a Server Action, this Server Action is no longer importable into Client Components. 

### Expected Behavior

Provided example works.

<sub>[NEXT-1227](https://linear.app/vercel/issue/NEXT-1227/server-only-is-not-compatible-with-server-actions)</sub>

**Comment by abuinitski** [comment]
@shuding That's my mistake for not setting this property. Actually, adding `experimental.serverActions: true` does not change the error.

**Comment by shuding** [comment]
@abuinitski No worries! I think the issue is that the Next.js version in that sandbox is a bit old, we don't have Server Actions implemented yet:

```
next@canary:
  version "13.1.7-canary.10"
```

Can you update it to the latest version? I tested locally and it works.

**Comment by zerosym** [comment]
Hi, I'm seeing the same error with the latest canary build. The exception occurs upon submission of the form / triggering the action.

**Comment by bradennapier** [comment]
Yes the issue occurs if anything that a server action imports has `server-only` in it - i have had to remove server only from my entire app

**Comment by shuding** [comment]
Should be fixed in #50257, thanks!
